vowpalwabbit/topk.h

```diff
@@ -3,4 +3,36 @@ Copyright (c) by respective owners including Yahoo!, Microsoft, and
 individual contributors. All rights reserved.  Released under a BSD
 license as described in the file LICENSE.
  */
+
+#pragma once
+
+#include "reductions.h"
+#include <queue>
+#include <utility>
+
 LEARNER::base_learner* topk_setup(VW::config::options_i& options, vw& all);
+
+namespace VW
+{
+  struct topk
+  {
+    using scored_example = std::pair<float, v_array<char>>;
+
+    topk(uint32_t K);
+
+    void predict(LEARNER::single_learner& base, multi_ex& ec_seq);
+    void learn(LEARNER::single_learner& base, multi_ex& ec_seq);
+    std::vector<scored_example> drain_queue();
```

Comment:
Should this be private?

Reply:
It doesn't seem right to link driver related output functions to the class itself. For the output function to work it needs access to the result of the reduction, hence why this function exists.